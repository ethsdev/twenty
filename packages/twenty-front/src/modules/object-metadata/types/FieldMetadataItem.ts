import { FieldDateMetadataSettings } from '@/object-record/record-field/types/FieldMetadata';

import { ThemeColor } from 'twenty-ui/theme';
import {
    Field,
    Object as MetadataObject,
    RelationDefinition,
    RelationMetadataType,
} from '~/generated-metadata/graphql';

export type FieldMetadataItemOption = {
  color: ThemeColor;
  id: string;
  label: string;
  position: number;
  value: string;
};

export type FieldMetadataItem = Omit<
  Field,
  '__typename' | 'defaultValue' | 'options' | 'relationDefinition'
> & {
  __typename?: string;
  defaultValue?: any;
  options?: FieldMetadataItemOption[] | null;
  relationDefinition?: {
    relationId: RelationDefinition['relationId'];
    direction: RelationMetadataType;
    sourceFieldMetadata: Pick<Field, 'id' | 'name'>;
    sourceObjectMetadata: Pick<
      MetadataObject,
      'id' | 'nameSingular' | 'namePlural'
    >;
    targetFieldMetadata: Pick<Field, 'id' | 'name'>;
    targetObjectMetadata: Pick<
      MetadataObject,
      'id' | 'nameSingular' | 'namePlural'
    >;
  } | null;
  settings?: FieldDateMetadataSettings;
  isLabelSyncedWithName?: boolean | null;
};
