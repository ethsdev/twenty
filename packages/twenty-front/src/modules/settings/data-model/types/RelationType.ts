import { RelationMetadataType } from '~/generated-metadata/graphql';

export type RelationType = RelationMetadataType;
