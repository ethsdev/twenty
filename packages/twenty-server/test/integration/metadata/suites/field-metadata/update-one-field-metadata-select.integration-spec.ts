import { createOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util';
import {
  LISTING_NAME_PLURAL,
  LISTING_NAME_SINGULAR,
} from 'test/integration/metadata/suites/object-metadata/constants/test-object-names.constant';
import { deleteOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { FieldMetadataType } from 'twenty-shared/types';

import { FieldMetadataComplexOption } from 'src/engine/metadata-modules/field-metadata/dtos/options.input';
import { UPDATE_CREATE_ONE_FIELD_METADATA_SELECT_TEST_CASES } from 'test/integration/metadata/suites/field-metadata/update-create-one-field-metadata-select-tests-cases';
import { updateOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/update-one-field-metadata.util';
import { createOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/create-one-object-metadata.util';

const { failingTestCases, successfulTestCases } =
  UPDATE_CREATE_ONE_FIELD_METADATA_SELECT_TEST_CASES;

describe('Field metadata select update tests group', () => {
  let createdObjectMetadataId: string;
  let createdFieldMetadata: string;

  beforeEach(async () => {
    const { data } = await createOneObjectMetadata({
      input: {
        labelSingular: LISTING_NAME_SINGULAR,
        labelPlural: LISTING_NAME_PLURAL,
        nameSingular: LISTING_NAME_SINGULAR,
        namePlural: LISTING_NAME_PLURAL,
        icon: 'IconBuildingSkyscraper',
        isLabelSyncedWithName: false,
      },
    });

    createdObjectMetadataId = data.createOneObject.id;

    const {
      data: { createOneField },
    } = await createOneFieldMetadata({
      input: {
        objectMetadataId: createdObjectMetadataId,
        type: FieldMetadataType.SELECT,
        name: 'testField',
        label: 'Test Field',
        isLabelSyncedWithName: false,
        options: [
          {
            label: 'Option 1',
            value: 'OPTION_1',
            color: 'green',
            position: 1,
          },
        ],
      },
      gqlFields: `
        id
        `,
    });
    createdFieldMetadata = createOneField.id;
  });

  afterEach(async () => {
    await deleteOneObjectMetadata({
      input: { idToDelete: createdObjectMetadataId },
    });
  });

  test.each(successfulTestCases)('$title', async ({ context: { options } }) => {
    const { data, errors } = await updateOneFieldMetadata({
      input: {
        idToUpdate: createdFieldMetadata,
        updatePayload: {
          options,
        },
      },
      gqlFields: `
        id
        options
        `,
    });

    expect(data.updateOneField).toBeDefined();
    const updatedOptions: FieldMetadataComplexOption[] =
      data.updateOneField.options;

    expect(errors).toBeUndefined();
    expect(updatedOptions.length).toBe(options.length);
    updatedOptions.forEach((option) => expect(option.id).toBeDefined());
    expect(updatedOptions).toMatchObject(options);
  });

  test.each(failingTestCases)('$title', async ({ context: { options } }) => {
    const { data, errors } = await updateOneFieldMetadata({
      input: {
        idToUpdate: createdFieldMetadata,
        updatePayload: {
          objectMetadataId: createdObjectMetadataId,
          name: 'testField',
          label: 'Test Field',
          isLabelSyncedWithName: false,
          options,
        },
      },
      gqlFields: `
        id
        options
        `,
    });

    expect(data).toBeUndefined();
    expect(errors).toBeDefined();
    expect(errors).toMatchSnapshot();
  });
});
