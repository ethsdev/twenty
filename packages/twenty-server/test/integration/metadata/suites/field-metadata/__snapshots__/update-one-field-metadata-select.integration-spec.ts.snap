// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field metadata select update tests group It should fail to create a field metada with an undefined option label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label is required",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metada with an undefined option value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value is required",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: empty string id  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: empty string label  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label "" is beneath 1 character",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: empty string value  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value "" is beneath 1 character",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: not a string id  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: not a string label  1`] = `
[
  {
    "extensions": {
      "code": "INTERNAL_SERVER_ERROR",
      "exceptionEventId": "mocked-exception-id",
    },
    "message": "label.includes is not a function",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: not a string value  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Value must be in UPPER_CASE and follow snake_case "22222"",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: null id  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is required",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: null label  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label is required",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: null value  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value is required",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: only white spaces id  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: only white spaces label  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label "" is beneath 1 character",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: only white spaces value  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value "" is beneath 1 character",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: too long id  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: too long label  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" exceeds 63 characters",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create a field metadata with an option having: too long value  1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" exceeds 63 characters",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create duplicated trimmed options values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Duplicated option value "OPTION_1"",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create field metadata with an option label that contains a comma 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Label must not contain a comma",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create field metadata with an option value that do not follow UPPER_CASE and snake_case 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Value must be in UPPER_CASE and follow snake_case "Option 1 and some other things, /"",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create field metadata with empty options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Options are required for enum fields",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create field metadata with null options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Options are required for enum fields",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create field metadata with options having duplicated values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Duplicated option value "OPTION_1"",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create field metadata with undefined options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Options are required for enum fields",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create option with invalid id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create two options with the same id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Duplicated option id "fd1f11fd-3f05-4a33-bddf-800c3412ce98"",
  },
]
`;

exports[`Field metadata select update tests group It should fail to create two options with the same position 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Duplicated option position "1"",
  },
]
`;
